### PR TITLE
Rewrite remote gdb packet handling

### PR DESCRIPTION
**Pull request by gitido**
This patch addresses several issues with the current implementation,
including:
- Excessive recursions
- Incorrect handling of the escape character
- Buffer size limitations

### CODE REVIEW

File: shlr/gdb/src/packet.c [hunk 1]
```diff
@@ -1,157 +1,181 @@
...
 
-void handle_chk(parsing_object_t* po) {
-	char checksum[3];
-	if (po->position >= po->length) return;
-	checksum[0] = get_next_token (po);
-	checksum[1] = get_next_token (po);
-	checksum[2] = '\0';
-	po->checksum = (uint8_t) strtol (checksum, NULL, 16);
-}
+static int append(libgdbr_t *g, char ch)
```
Comment (ret2libc) [inline shlr/gdb/src/packet.c]: We usually put the open brace on the same line as the function

File: shlr/gdb/src/packet.c [hunk 2]
```diff
@@ -1,157 +1,181 @@
...
-	token = get_next_token (po);
-	if (token == '#') {
-		po->end = po->position - 1; // subtract 2 cause of # itself and the incremented position after getNextToken
-		handle_chk (po);
-	} else if (token == '{') {
-		handle_escape (po);
-	} else handle_data (po);
-}
+	if (g->data_len == g->data_max) {
+		ptr = realloc(g->data, g->data_max * 2);
```
Comment (ret2libc) [inline shlr/gdb/src/packet.c]: Function calls should have a space between the name and the arguments

File: shlr/gdb/src/packet.c [hunk 3]
```diff
@@ -1,157 +1,181 @@
...
-				eprintf ("Runlength decoding error\n");
+	int j = 0;
+
+	for (i = 0; i < len; i++) {
+		char cur = g->read_buff[i];
+
+		if (ctx->flags & CHKSUM) {
+			ctx->chksum -= hex2int(cur) << (ctx->chksum_nibble * 4);
+
+			if (!--ctx->chksum_nibble)
```
Comment (ret2libc) [inline shlr/gdb/src/packet.c]: If the 'if' statement is on two lines, like this case, you should wrap it with braces.

**Comment by ret2libc** [comment]
Thanks for the contribution! Please have a quick look at https://github.com/radare/radare2/blob/master/CONTRIBUTING.md .
In particular, I commented in the PR some of the issues I've seen, but there are others. Please pay attention to:
* open brace for functions should be on the same line and there shouldn't be any space between the function name and parameters in the signature.
* if statement should be wrap with braces
* when you call a function, we use a space between the function name and the actual arguments.

**Comment by gitido** [comment]
Thanks for reviewing this!

This generally follows the Linux kernel coding convention, but sure, I can modify this patch to match the rest of the codebase.

Other than white spaces, any other issues with this patch?

**Comment by ret2libc** [comment]
Yes I know but we don't follow linux kernel coding style unfortunately :( 
Unfortunately I'm not familiar at all with GDB stuff, so I'll let someone else review in a better way this patch! Thanks again ;)

**Comment by XVilka** [comment]
Looks good to me. I'll check it today's evening more thoroughly.

**Comment by XVilka** [comment]
@gitido can you please add reading 'PacketSize’ from the '‘qSupported' request? And using it intead of hardcoded value?

**Comment by gitido** [comment]
@ret2libc Pushed a new version with adjusted coding style
@XVilka Thanks for reviewing this!

**Comment by gitido** [comment]
@Xvilka I guess that should be handled by a separate patch, since this code only deals with packets unmarshalling and not actual contents.

By the way, the initial motivation for this rewrite was that previously r2 crashed quite often when escaping is used (say, when debugging x86_64 targets).

**Comment by XVilka** [comment]
@gitido yes, you right about spitted patch. You can put a few patches at once in the PR though.

**Comment by gitido** [comment]
@XVilka In any case, I believe it is now less important than it used to - with this implementation we do support larger packet sizes than the hard coded value (notice the realloc there). That wasn't the case before.

**Comment by radare** [comment]
Just tested against lldb's debugserver and its segfaulting hard, in several places. so its buggy, not merging

**Comment by radare** [comment]
this fixes the hang after `dc` when the process dies. and no segfaults at all, sorry for the wrong test

**Comment by radare** [comment]
thanks!

**Pull request closed without merging**